### PR TITLE
Documentation for Combinatorial Optimization

### PR DESCRIPTION
**Pull request by DangGiaChi**
* Documentation for Combinatorial Optimization
* Added Splits in Algorithmic Reasoning so that it is consistent with the structure of BlueSky and CO

### CODE REVIEW

File: graphbench/datasets/_combinatorial_optimization.py [hunk 1]
```diff
@@ -39,6 +39,80 @@
...
+
+    Overview:
+        On the CO datasets, we consider 3 classic NP-hard graph problems: the maximum independent set (MIS), the max-cut, and the graph coloring problem. 
+        However, the provided datasets currently focus on the MIS problem, with max-cut and graph coloring planned for future work.
+
+        We synthetically generate optimization problems across 3 well-established random graph families:
+
+        - Barabási-Albert (BA)
+        - Erdős-Rényi (ER)
+        - Regular Bipartite (RB)
```
Comment (ArmanMielke) [inline graphbench/datasets/_combinatorial_optimization.py]: Where does the name "Regular Bipartite" come from? As far as I'm aware the original paper just calls it model RB: https://www.sciencedirect.com/science/article/pii/S0004370207000653 In the graph learning literature they're called "RB graphs" to be consistent with "ER graphs" etc., though I'd also be interested in knowing whether there's a long version of the name. In any case, checking the graphs in the dataset shows that they're not bipartite.

File: graphbench/datasets/_combinatorial_optimization.py [hunk 2]
```diff
@@ -39,6 +39,80 @@
...
+
+        Each graph family is available in 2 configurations: small and large, totaling 6 distinct datasets.
```
Comment (ArmanMielke) [inline graphbench/datasets/_combinatorial_optimization.py:91]: I think we should state the total number of graphs in each dataset (50,000), as well as the graph sizes for small and large graphs. Those are probably the most important numbers for the user. It's probably also worth mentioning that the BA graphs are considerably less dense than the other two with the parameters we are using (I know you had no way of knowing that, but might be useful information ^^').

File: graphbench/datasets/_combinatorial_optimization.py [hunk 3]
```diff
@@ -39,6 +39,80 @@
...
+        where generator is one of ``ba``, ``er``, or ``rb``, and size is either ``small`` or ``large``.
+
+        Hence, valid dataset names for the loader are:
+        ``co_ba_small``,
+        ``co_er_small``,
+        ``co_rb_small``,
+        ``co_ba_large``,
+        ``co_er_large``,
+        and ``co_rb_large``.
+
```
Comment (ArmanMielke) [inline graphbench/datasets/_combinatorial_optimization.py:128]: Add

> In addition to this, we provide ``co`` as a convenience identifier to load all of the above datasets.

I know this can be easy to miss in `datasets.csv` because that is not in any intuitive order... A lot of the other domains also have convenience aliases like this.

File: graphbench/datasets/_combinatorial_optimization.py [hunk 4]
```diff
@@ -39,6 +39,80 @@
...
+           :header-rows: 1
+
+           * - Task
+             - Node features
+             - Edge features
+             - Target
+           * - Maximum Independent Set (MIS)
+             - None
+             - None
+             - Optimal objective value of the given CO instance, size ``[1]``
```
Comment (ArmanMielke) [inline graphbench/datasets/_combinatorial_optimization.py]: Please state exactly what the fields are called; the user needs that to use the dataset. You can load up one of the datasets in an interactive Python shell to inspect a graph (BA small is the smallest one, so should download quickly). You don't need to mention `edge_index` and `num_nodes`, since every graph has that.

You'll find that we also provide the actual MIS solution on top of the objective value in case someone wants to use that. Because there can be multiple fields per task, I think this table works better with attributes as rows instead of tasks, and "Attribute name", "Size", and "Description" as the columns. You can take a look at the Bluesky docs for an example of that.

File: graphbench/datasets/_combinatorial_optimization.py [hunk 5]
```diff
@@ -55,18 +129,16 @@ def __init__(
...
+        Args:
+            name: Dataset identifier such as 'co_ba_small', 'co_er_large', etc.
+            split: One of 'train', 'val', 'test'.
+            root: Root directory where the dataset folder will be created.
+            transform, pre_transform: Optional PyG transforms applied at load time.
+            target: Optional task variant (unused for unsupervised tasks).
+            generate: If True, generate synthetic graphs instead of downloading.
+            num_samples: Number of synthetic graphs to generate when generate=True.
+            cleanup_raw: Whether to remove raw files after processing.
+            load_preprocessed: If True, load existing processed objects instead of regenerating.
```
Comment (ArmanMielke) [inline graphbench/datasets/_combinatorial_optimization.py]:
```suggestion
            name: Dataset identifier in the form ``co_{graph_type}_{size}``, e.g. ``co_rb_large``.
            split: Whether to load the train, validation, or test split of the dataset.
            root: Root directory where the dataset folder will be created.
            transform: Optional PyG transform applied to data objects before every access.
            pre_transform: Optional PyG transform applied before saving data objects to disk.
            pre_filter: A function that indicates whether a data object should be included in the final dataset.
            target: Optional task variant (unused for unsupervised tasks).
            generate: If True, generate synthetic graphs instead of downloading.
            num_samples: Number of synthetic graphs to generate when generate=True.
            cleanup_raw: If True, remove raw files after processing.
            load_preprocessed: If True, load existing processed objects instead of regenerating.
```

Comment (ArmanMielke) [reply graphbench/datasets/_combinatorial_optimization.py]: If you generate the documentation, you'll see that it shows "train", "val", "test" as the options for `split` right next to the description, so saying that again is a bit redundant. You can honestly just copy almost all of the parameter descriptions from AlgoReasDataset or BlueSkyDataset (did that here because I wanted to try out the suggest change feature ^^).

Also, make sure to document all parameters.

Comment (ArmanMielke) [reply graphbench/datasets/_combinatorial_optimization.py]: I also just noticed that the `target` parameter is completely unused, probably best to delete that (not sure who put that here in the first place)

File: graphbench/datasets/_combinatorial_optimization.py [hunk 6]
```diff
@@ -39,6 +39,80 @@
...
+    Combinatorial Optimization (CO) datasets.
+
+    Note:
+        This class **should only be used directly when generating new datasets**.
+        To access provided datasets, please consider using :class:`graphbench.Loader`.
+        The sections below give details on the data available through the :class:`graphbench.Loader` interface.
+
+    Overview:
+        On the CO datasets, we consider 3 classic NP-hard graph problems: the maximum independent set (MIS), the max-cut, and the graph coloring problem. 
+        However, the provided datasets currently focus on the MIS problem, with max-cut and graph coloring planned for future work.
```
Comment (ArmanMielke) [inline graphbench/datasets/_combinatorial_optimization.py]: The thing that is missing for maximum cut and graph coloring is only the ground truth labels for the supervised setting. The unsupervised setting can be done on all tasks because it doesn't need training labels. I should probably have cleared that up in my message the other day, sorry about that.

I think it's worth spending at least a couple sentences to describe how the supervised and unsupervised setups work, since that information is necessary to use the datasets.

Somewhat unrelated to this PR, we have helper functions for unsupervised learning in the other repository. They're not strictly necessary to use the datasets for unsupervised CO, but they do take a lot of work out of the user's hands, so it would be nice to provide them (we also say something to that effect in the paper). Can you copy those over and find a nice place for them in the API? Timo can give you access to the other repo if you don't have it already. The relevant files are in tasks/co/, there you want all the functions in decoders.py, unsupervised_losses.py and validate_solution.py. In terms of documentation they don't need much. E.g. for the loss functions, it can just be "Unsupervised loss function to train a model for [task name]. \n\n At test time, use [link to the decoder] to convert the model's soft output to a discrete solution, and [link to mis_size/max_cut_size/...] to evaluate the model's performance." plus parameter descriptions of course. For the decoders, you can use "Converts the model's soft prediction to a discrete solution to the [task name] problem. \n\n This can be used at test time for models trained with [link to the loss function].", again plus parameter descriptions.

File: graphbench/datasets/_combinatorial_optimization.py [hunk 7]
```diff
@@ -39,6 +39,80 @@
...
+        Each graph family is available in 2 configurations: small and large, totaling 6 distinct datasets.
+
+        Please refer to the `GraphBench paper <https://arxiv.org/abs/2512.04475>`__ for the exact parameters used for graph generation.
+
+    
+    Splits:
+        By default, pre-generated datasets are already split into training, validation, and test sets.
+        
+        When generating synthetic datasets dynamically, an automatic split of 70% training, 15% validation,
+        and 15% testing is applied.
```
Comment (ArmanMielke) [inline graphbench/datasets/_combinatorial_optimization.py]: It's always 70/15/15, so no need to mention pre-generated datasets and newly generated datasets separately. When documenting the other datasets, you can load them in an interactive Python shell to get a sense of what the splits look like (+ the paper usually also mentions it).

**Review by ArmanMielke — CHANGES_REQUESTED** [review]
The information that is there is mostly good, but there's also a lot of information missing. I recommend always loading the datasets to inspect them.

File: graphbench/datasets/_combinatorial_optimization.py (hunk 6 above)
Comment (ArmanMielke) [reply graphbench/datasets/_combinatorial_optimization.py]: If you add those (in a separate PR), it's worth linking to it somewhere in the CODataset doc.

File: graphbench/datasets/_combinatorial_optimization.py (hunk 1 above)
Comment (DangGiaChi) [reply graphbench/datasets/_combinatorial_optimization.py]: This was actually my oversight. I was curious and tried to see if RB stands for anything and found papers on Random Regular Bipartite Graphs. But after reading carefully they are indeed completely different things.

I'll make sure to research more carefully next time!

**Review by ArmanMielke — APPROVED** [review]
Looks good now! I'll make some adjustments on the descriptions of the supervised/unsupervised setups, but let's merge first.
